codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolUtils.java

```diff
@@ -129,51 +128,86 @@ public static StructureShape expectOutput(Model model, OperationShape operation)
-        String operand = container + "." + memberName;
+        Shape targetShape = context.getModel().expectShape(member.getTarget());
+        Shape container = context.getModel().expectShape(member.getContainer());
 
-        boolean enumShape = target.hasTrait(EnumTrait.class);
+        boolean withAddr = !GoPointableIndex.of(context.getModel()).isPointable(member)
+                && GoPointableIndex.of(context.getModel()).isPointable(targetShape);
+        boolean isMap = container.getType() == ShapeType.MAP;
 
-        if (!enumShape && !CodegenUtils.isNilAssignableToShape(model, member)) {
-            consumer.accept(operand);
-            return;
+        String var = origDestVar;
+        if (isMap) {
+            writer.write("mapVar := $L", origDestVar);
+            var = "mapVar";
         }
 
-        String conditionCheck;
-        if (enumShape) {
-            conditionCheck = "len(" + operand + ") > 0";
-        } else {
-            conditionCheck = operand + " != nil";
+        if (withAddr) {
+            writer.write("destAddr := &$L", var);
+            var = "destAddr";
         }
 
-        context.getWriter().openBlock("if $L {", "}", conditionCheck, () -> {
-            consumer.accept(operand);
-        });
+        lambda.accept(var);
+
+        if (isMap || withAddr) {
+            if (withAddr) {
+                var = "*" + var;
+            }
+
+            writer.write("$L = $L", origDestVar, var);
+        }
     }
 
     /**
-     * Determines whether a given shape will use a scalar when the shape is used as a union value.
+     * Writes helper variables for the delegation function to ensure that map values are safely delegated down
+     * each level.
      *
-     * @param shape the shape to check
-     * @return false if the shape should use pointers
+     * @param context     generation context
+     * @param writer      go writer
+     * @param member      shape to determine if pointable
+     * @param origDestVar original variable name
+     * @param lambda      runnable
      */
-    public static boolean usesScalarWhenUnionValue(Shape shape) {
-        return !(shape instanceof SimpleShape) || shape.isBlobShape() || shape.hasTrait(EnumTrait.class)
-                || shape.hasTrait(StreamingTrait.class);
+    public static void writeSerDelegateFunction(
+            GenerationContext context,
+            GoWriter writer,
+            MemberShape member,
+            String origDestVar,
+            Consumer<String> lambda
+    ) {
+        Shape targetShape = context.getModel().expectShape(member.getTarget());
+        Shape container = context.getModel().expectShape(member.getContainer());
+
+        boolean withAddr = !GoPointableIndex.of(context.getModel()).isPointable(member)
+                && GoPointableIndex.of(context.getModel()).isPointable(targetShape);
+        boolean isMap = container.getType() == ShapeType.MAP;
+
+        String var = origDestVar;
```

Comment:
Minor nit: Later versions of Java can use `var` as a type identifier similar to `:=` in Go without having to specify the actual type. That is why it is getting syntax highlighted here in GitHub. I would argue the intention would be clearer if we chose an alternative rather then `var` for a variable. Similar reasoning to why I think `obj` is a poor variable name in Java.

Reply:
did not know that, can rename this to something that won't clash.